source/common/sink_impl.cc

```diff
@@ -0,0 +1,109 @@
+#include "common/sink_impl.h"
+
+#include <array>
+#include <filesystem>
+#include <fstream>
+
+#include "external/envoy/source/common/common/logger.h"
+#include "external/envoy/source/common/common/random_generator.h"
+
+namespace Nighthawk {
+namespace {
+
+absl::Status verifyStringIsGuid(absl::string_view s) {
+  Envoy::Random::RandomGeneratorImpl random;
+  const std::string reference_value = random.uuid();
+  const std::string err_template = "'{}' is not a guid: {}";
+
+  if (s.size() != reference_value.size()) {
+    return absl::Status(absl::StatusCode::kInvalidArgument,
+                        fmt::format(err_template, s, "bad string length."));
+  }
+  for (size_t i = 0; i < s.size(); i++) {
+    if (reference_value[i] == '-') {
+      if (s[i] != '-') {
+        return absl::Status(
+            absl::StatusCode::kInvalidArgument,
+            fmt::format(err_template, s, "expectations around '-' positions not met."));
+      }
+      continue;
+    }
+    if (!std::isxdigit(s[i])) {
+      return absl::Status(absl::StatusCode::kInvalidArgument,
+                          fmt::format(err_template, s, "unexpected character encountered."));
+    }
+  }
+  return absl::OkStatus();
+}
+
+} // namespace
+
+absl::Status FileSinkImpl::StoreExecutionResultPiece(
+    const ::nighthawk::client::ExecutionResponse& response) const {
+  const std::string& execution_id = response.execution_id();
+  absl::Status status = verifyStringIsGuid(execution_id);
+  if (!status.ok()) {
+    return status;
+  }
+  std::error_code error_code;
+  std::filesystem::create_directories("/tmp/nh/" + std::string(execution_id) + "/", error_code);
+  // Note error_code will not be set if an existing directory existed.
+  if (error_code.value()) {
+    return absl::Status(absl::StatusCode::kInternal, error_code.message());
+  }
+  std::array<char, L_tmpnam> name_buffer;
+  // This emits a compiler warning: the use of `tmpnam' is dangerous, better use `mkstemp'.
```

Comment:
This begs a question as to why we don't use the suggested `mkstemp`. If we do have a good reason, can we expand the comment to include that information?

Reply:
Rewritten en 8f8a5da to not rely on mkstemp